### PR TITLE
redux-promise-middleware: fix Dispatch return type

### PR DESCRIPTION
**Pull request by madprog**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/17598
- [x] Increase the version number in the header if appropriate.
- ~If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.~

### CODE REVIEW

File: types/redux-promise-middleware/index.d.ts [hunk 1]
```diff
@@ -2,7 +2,23 @@
...
 // Definitions by: ianks <https://github.com/ianks>
 // Definitions: https://github.com/borisyankov/DefinitelyTyped
+// TypeScript Version: 2.8
 
 import { Middleware } from 'redux';
 
 export default function promiseMiddleware(config?: { promiseTypeSuffixes?: string[], promiseTypeDelimiter?: string }): Middleware;
+
+declare module "redux" {
+  export interface PromiseAction<R> extends Action {
```
Comment (sandersn) [inline types/redux-promise-middleware/index.d.ts]: Does redux-promise-middleware make this change to `Dispatch`? Or should this change go into redux' types?

**Review by sandersn — COMMENTED** [review]
Basic question since I don't know much about redux or redux-promise-middleware.

File: types/redux-promise-middleware/index.d.ts (hunk 1 above)
Comment (madprog) [reply types/redux-promise-middleware/index.d.ts]: Hello @sandersn,

From my understanding of these libraries, the promise middleware will return [a promise object](https://github.com/pburtchaell/redux-promise-middleware/blob/master/src/index.js#L203) whereas redux will [return the action](https://github.com/reduxjs/redux/blob/master/src/createStore.js#L197) instead. So yes, I believe that this change should be here, as an override of `Dispatch`. I was inspired from what was done in [redux-thunk 2.2.0](https://github.com/reduxjs/redux-thunk/blob/v2.2.0/index.d.ts#L8) (another middleware for Redux, which makes dispatch accept a function as well as an action).

However, I have just noticed that the last version (2.3.0) of redux-thunk has [changed their type definition](https://github.com/reduxjs/redux-thunk/compare/v2.2.0...v2.3.0#diff-b52768974e6bc0faccb7d4b75b162c99), but I am not sure that understand how it works after their changes.

Comment (sandersn) [reply types/redux-promise-middleware/index.d.ts]: I don’t know anything about redux, but from redux-thunk’s tests, it looks like redux has a new function called applyMiddleware that lets you avoid monkeypatching redux objects. I’m not sure if redux-promise-middleware uses this function yet.

Comment (sandersn) [reply types/redux-promise-middleware/index.d.ts]: If it doesn’t, then your correct change is probably correct.

Comment (madprog) [reply types/redux-promise-middleware/index.d.ts]: This is not really about monkeypatching redux objects.
In the redux world, a middleware is a function that returns a `Dispatch` function. These functions are chained together, so each middleware's `Dispatch` receives the state of the previous one, and pass it to the next, until redux's dispatch is called.
Usually, a middleware will just return the same value as its underlying function.
However, redux-promise-middleware is returning a promise object instead, and this PR is trying to reflect this.

Comment (sandersn) [reply types/redux-promise-middleware/index.d.ts]: If you're not mutating the state, imitating the new version of redux-thunk is probably the right answer. It looks like `Middleware` has a bunch of type parameters that allow you to specify your custom Dispatch function that way.

**Pull request merged**